instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshOnDemandTest.kt

```diff
@@ -174,6 +178,193 @@ class RouteRefreshOnDemandTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::
         )
     }
 
+    @Test
+    fun route_refresh_on_demand_after_alternatives_change_between_attempts() = sdkTest {
+        createMapboxNavigation(createRouteRefreshOptionsWithInvalidInterval(5_000))
+        mockWebServerRule.requestHandlers.add(
+            MockDirectionsRequestHandler(
+                "driving-traffic",
+                readRawFileText(activity, R.raw.route_response_route_refresh),
+                twoCoordinates
+            )
+        )
+        val refreshHandler = MockDirectionsRefreshHandler(
+            "route_response_route_refresh",
+            readRawFileText(activity, R.raw.route_response_route_refresh_annotations),
+        )
+        refreshHandler.jsonResponseModifier = DynamicResponseModifier()
+        mockWebServerRule.requestHandlers.add(NthAttemptHandler(refreshHandler, 1))
+        val observer = TestObserver()
+
+        mapboxNavigation.routeRefreshController.registerRouteRefreshStateObserver(observer)
+        mapboxNavigation.startTripSession()
+        val routeOptions = generateRouteOptions(twoCoordinates)
+        val requestedRoutes = mapboxNavigation.requestRoutes(routeOptions)
+            .getSuccessfulResultOrThrowException()
+            .routes
+        val initialRoutes = requestedRoutes.take(1)
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(initialRoutes)
+
+        delay(7000) // in-between attempts
+
+        // add alternatives
+        mapboxNavigation.setNavigationRoutesAndWaitForAlternativesUpdate(requestedRoutes)
+
+        val refreshes = mutableListOf<RoutesUpdatedResult>()
+        mapboxNavigation.registerRoutesObserver {
+            if (it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH &&
+                it.navigationRoutes.map { it.id } == requestedRoutes.map { it.id }) {
+                refreshes.add(it)
+            }
+        }
+        mapboxNavigation.routeRefreshController.requestImmediateRouteRefresh()
+
+        delay(4000)
+        assertEquals(1, refreshes.size) // 1 refresh from refresh on demand
+
+        withTimeout(2500) { // 5s interval, 1.5s accuracy
+            while (refreshes.size < 2) {
+                delay(100)
+            }
+        }
+        assertEquals(2, refreshes.size)
+
+        assertEquals(
+            listOf(
+                RouteRefreshExtra.REFRESH_STATE_STARTED,
+                RouteRefreshExtra.REFRESH_STATE_CANCELED,
+                RouteRefreshExtra.REFRESH_STATE_STARTED,
+                RouteRefreshExtra.REFRESH_STATE_FINISHED_SUCCESS,
+                RouteRefreshExtra.REFRESH_STATE_STARTED,
+                RouteRefreshExtra.REFRESH_STATE_FINISHED_SUCCESS,
+            ),
+            observer.getStatesSnapshot()
+        )
+    }
+
+    @Test
+    fun route_refresh_on_demand_after_reroute() = sdkTest {
```

Comment:
have you considered verifying behaviour in a unit test which tests `RouteRefreshController` which uses real implementation or Immediate and planned route request so that only router is mocked? I think those types of situations could be emulated there.

Reply:
Yes. We have such unit tests. But the point here is that routes change externally (from MapboxNavigation). There is also filtering by reason there. That's why I want to check everything together here.

Reply:
> There is also filtering by reason there.

Why not to move filtering by reason to route refresh controller?

Reply:
I updated the tests.